### PR TITLE
Qualcomm AI Engine Direct - Quantizer refine for qat

### PR DESCRIPTION
**Pull request by chunit-quic**
- Reorginize qualcomm/quantizer
- Split quantizer/utils.py to 
-- qconfig
-- annotators
-- observers directory
- Change coresponding callees
- Rename get_default_Nbit_qnn_ptq_config to get_NaNw_qnn_ptq_config
- Add 16a4w conv test* (It is not compared with original model)

### CODE REVIEW

**Comment by chunit-quic** [comment]
Hi @cccclai,

We implement the draft PR for
1. Refine API of QnnQauntizer
2. Add more qat quant confings
3. Add 16a4w qat conv test case

We have another local branch which is testing QAT mobilenet v2 based on it. Yet we encounter some problems. 
During QAT process, the the accuracy/loss seems to be reasonable if we only quantize conv ops. Yet if we quantize any other op, accuracy/loss drop significantly. Is there any known issue for this? Or maybe it just results from some missetting form us. (We can make a patch or update the local branch here if you are insterested in it)

 Thank you. :)

**Comment by cccclai** [comment]
@navsud can you review this PR?

File: backends/qualcomm/quantizer/observers/param_observer.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
+import torch
+from torch.ao.quantization.observer import UniformQuantizationObserverBase
+
+
+class ParamObserver(UniformQuantizationObserverBase):
```
Comment (navsud) [inline backends/qualcomm/quantizer/observers/param_observer.py:5]: This looks great and is generic enough to be added into torch/ao/quantization/observer.py. If possible, please move it there as a follow-up. 

How about renaming it to PerChannelParamObserver() or PerChannelWeightObserver?

File: backends/qualcomm/quantizer/qconfig.py [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+    )
+
+    weight_fake_quant_ctr = FusedMovingAvgObsFakeQuantize.with_args(
+        dtype=torch.int8,
+        quant_min=-7,
+        quant_max=7,
+        qscheme=torch.per_tensor_symmetric,
+        ch_axis=0,
+        reduce_range=True,
+        observer=MovingAverageMinMaxObserver,
```
Comment (navsud) [inline backends/qualcomm/quantizer/qconfig.py:362]: This should be PerChannelMovingAverageMinMaxObserver and qscheme=torch.per_channel_symmetric?

File: backends/qualcomm/quantizer/quantizer.py [hunk 3]
```diff
@@ -22,25 +22,29 @@
...
+from .qconfig import (
+    get_16a16w_qnn_ptq_config,
     get_16a4w_qnn_ptq_config,
+    get_16a4w_qnn_qat_config,
     get_16a8w_qnn_ptq_config,
-    get_default_16bit_qnn_ptq_config,
-    get_default_8bit_qat_proto,
-    get_default_8bit_qnn_ptq_config,
+    get_8a8w_qnn_ptq_config,
+    get_8a8w_qnn_qat_config,
```
Comment (navsud) [inline backends/qualcomm/quantizer/quantizer.py:31]: As a follow-up PR, can we unify/simplify: `get_8a8w_qnn_ptq_config()` and `get_8a8w_qnn_qat_config()` into `get_8a8w_qnn_config()` with the argument is_train=True/False, which will define whether we are doing PTQ vs. QAT.

**Review by navsud — COMMENTED** [review]
Overall LGTM.
cc: @cccclai for final approval.

File: backends/qualcomm/quantizer/observers/param_observer.py (hunk 1 above)
Comment (chunit-quic) [reply backends/qualcomm/quantizer/observers/param_observer.py]: It's great to see you like it! Change the name and add a TODO for it

File: backends/qualcomm/quantizer/qconfig.py (hunk 2 above)
Comment (chunit-quic) [reply backends/qualcomm/quantizer/qconfig.py:363]: Sorry for misunderstanding. This quant config is for per tensor specifically. The per channel one is [here](https://github.com/pytorch/executorch/pull/6513/files#diff-c990fa4562a79a98b35cc2429e09dc2461e175cc7bf0a8b112655fef021dc5edR400), and we use[ the condition of a quantizer member function here](https://github.com/pytorch/executorch/pull/6513/files#diff-4cb9d05b0ef6711a9bc59c1556fd35ee441ece67389052a740117f9d1cd04a0cR122) to assign per channel quant confing.

File: backends/qualcomm/quantizer/quantizer.py (hunk 3 above)
Comment (chunit-quic) [reply backends/qualcomm/quantizer/quantizer.py:33]: No problem. We would like to keep separated at first. Once it seems to be stable we will raise a TODO PR.

**Comment by chunit-quic** [comment]
> Overall LGTM. cc: @cccclai for final approval.

Hi @navsud

Thank you very much for reviewing! We just uploaded a patch based on comments. Just two questions we would like to ask.

1. Does QAT only work for ops with weight at this moment? I tried to do QAT for a single op add, and got complained about no parameters to be trained. Is it expected?
2. Similar to the point 1. We are not pretty sure whether we quantize ops of mobilenetV2 properly. Accuracy drop significantly during QAT if we add quant config to any op but convolution op. More details can be found [in the first comment of the PR](https://github.com/pytorch/executorch/pull/6513#issuecomment-2440510907)

Thank you for reading! If anything is unclear please feel free to let me know. I will try to describe it more. :D

**Review by navsud — APPROVED** [review]
LGTM

**Comment by chunit-quic** [comment]
Hi @cccclai, may you take a glance at this PR when you are free, and perhas merge it, if it looks good to you? Thanks :D

**Review by cccclai — APPROVED** [review]
Looks good to me - thank you for adding the feature!

**Comment by cccclai** [comment]
I notice that I didn't import to internal...in the worst case, if this diff breaks some tests, we may need to revert and reland it...

**Comment by chunit-quic** [comment]
> I notice that I didn't import to internal...in the worst case, if this diff breaks some tests, we may need to revert and reland it...

It's fine. If so just let me know what should I fix and I will do it asap.  :)

**Comment by cccclai** [comment]
@chunit-quic Unfortunately it's reverted because it breaks internal test....can you submit a PR again?

**Comment by chunit-quic** [comment]
> @chunit-quic Unfortunately it's reverted because it breaks internal test....can you submit a PR again?

No problem. The new PR is [PR6747](https://github.com/pytorch/executorch/pull/6747)
